vllm/model_executor/guided_decoding/xgrammar_decoding.py

```diff
@@ -37,11 +38,21 @@ def get_local_xgrammar_guided_decoding_logits_processor(
     return XGrammarLogitsProcessor(config)
 
 
-class TokenizerData(NamedTuple):
+@dataclass(frozen=True)
+class TokenizerData:
     """Immutable container for cached tokenizer data."""
-    encoded_vocab: list[str]
-    stop_token_ids: list[int] | None
-    backend_str: str
+    encoded_vocab: list[str] = field(default_factory=list)
+    stop_token_ids: list[int] | None = None
+    # These fields are mutually exclusive: `backend_str` is used to create a
+    # TokenizeInfo with `TokenizerInfo.from_huggingface` while `vocab_type` is
+    # used within the constructor of TokenizeInfo
+    backend_str: str | None = None
+    vocab_type: xgr.VocabType | None = None
+
+    def __post_init__(self):
+        # Check for mutual exclusive
+        assert not (self.backend_str and self.vocab_type), \
+            "backend_str and vocab_type are mutual exclusive"
```

Comment:
hmm, is it possible to just either use backend_str or vocab_type here?

the cartesian product is probably harder to document.

Reply:
> hmm, is it possible to just either use backend_str or vocab_type here?

The way `get_tokenizer_data` was implemented, I don't see another way. The method `from_hugging_face` from xgrammar have several paths, and depending the type of the tokenizer it needs backend_str or vocab_type. 

> the cartesian product is probably harder to document.

I'm sorry, what you mean by cartesian product? Can you elaborate more?

Reply:
There are two mutual exclusive options, which will have different combinations across different multiples.

Considering how much models vLLM is currently supported, in this case, we will have to draw the table of what combinations work/doesn't work, hence the cartesian product of combinations.

Reply:
Ok got it. Do you think documentation is required for this PR? My primary goal was to address the issue with guided decoding when using tokenizer_mode mistral, as mentioned in the related [PR](https://github.com/vllm-project/vllm/issues/11045). Without this fix, vLLM crashes for xgrammar due to an unhandled code path.

In my opinion, it might be beneficial to apply these fixes and then assess whether extensive documentation or code checks are required to help users understand unsupported features. What do you think?